File: backend/src/app/api/v1/collaboration.py
"""Collaborative Editing API Endpoints.

Provides REST and WebSocket endpoints for real-time collaborative
gallery editing features including sessions, presence, and actions.
"""

from __future__ import annotations

import asyncio
import json
import logging
from typing import Annotated, Any
from uuid import UUID

from fastapi import (
    APIRouter,
    Depends,
    HTTPException,
    Query,
    WebSocket,
    WebSocketDisconnect,
    status,
)
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.dependencies.auth import get_current_user
from app.api.dependencies.db import get_db
from app.db.redis import get_redis_client
from app.services.collaboration_service import (
    CollaborationService,
    get_collaboration_service,
    COLLAB_CHANNEL_PREFIX,
)

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/collaboration", tags=["collaboration"])


# ---------------------------------------------------------------------------
# Request/Response Models
# ---------------------------------------------------------------------------

class CreateSessionRequest(BaseModel):
    """Request to create or join an edit session."""
    gallery_id: UUID = Field(..., description="Gallery to edit collaboratively")


class CreateSessionResponse(BaseModel):
    """Response from creating/joining a session."""
    session_id: str
    gallery_id: str
    workspace_id: str
    version: int
    collaborators: list[dict[str, Any]]
    your_color: str
    websocket_url: str
    token: str


class UpdatePresenceRequest(BaseModel):
    """Request to update presence information."""
    status: str | None = Field(None, description="Status: active, idle")
    cursor: dict[str, Any] | None = Field(None, description="Cursor position")
    selected_assets: list[str] | None = Field(None, description="Selected asset IDs")
    focus_area: str | None = Field(None, description="Current focus area")


class RecordActionRequest(BaseModel):
    """Request to record a collaborative action."""
    action_type: str = Field(..., description="Type of action")
    payload: dict[str, Any] = Field(default_factory=dict, description="Action payload")
    base_version: int = Field(..., description="Version this action is based on")


class RecordActionResponse(BaseModel):
    """Response from recording an action."""
    action_id: str
    success: bool
    new_version: int
    conflict: dict[str, Any] | None = None


class AcquireLockRequest(BaseModel):
    """Request to acquire a resource lock."""
    resource_type: str = Field(..., description="Type: gallery, sub_gallery, asset, settings")
    resource_id: UUID = Field(..., description="Resource UUID")
    lock_type: str = Field(default="exclusive", description="Lock type: exclusive, shared")
    duration_seconds: int = Field(default=300, description="Lock duration in seconds")
    reason: str | None = Field(None, description="Reason for the lock")


class LockResponse(BaseModel):
    """Response from lock operations."""
    success: bool
    lock: dict[str, Any] | None = None
    error: str | None = None
    holder: dict[str, str] | None = None


class ActionHistoryResponse(BaseModel):
    """Response with action history."""
    actions: list[dict[str, Any]]
    from_version: int
    to_version: int


class UndoActionRequest(BaseModel):
    """Request to undo an action."""
    action_id: UUID = Field(..., description="Action to undo")


# ---------------------------------------------------------------------------
# REST Endpoints
# ---------------------------------------------------------------------------

@router.post(
    "/sessions",
    response_model=CreateSessionResponse,
    summary="Create or join collaborative editing session",
)
async def create_session(
    request: CreateSessionRequest,
    current_user: dict = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> CreateSessionResponse:
    """Create a new collaborative editing session or join an existing one.

    If an active session already exists for the gallery, the user will join it.
    Otherwise, a new session is created.

    Returns session info including WebSocket URL and authentication token.
    """
    user_id = UUID(current_user["user_id"])
    workspace_id = UUID(current_user["workspace_id"])

    service = get_collaboration_service(db)
    result = await service.create_session(
        workspace_id=workspace_id,
        gallery_id=request.gallery_id,
        user_id=user_id,
    )

    session = result["session"]
    return CreateSessionResponse(
        session_id=session["session_id"],
        gallery_id=session["gallery_id"],
        workspace_id=session["workspace_id"],
        version=session["version"],
        collaborators=session["collaborators"],
        your_color=result["your_color"],
        websocket_url=result["websocket_url"],
        token=result["token"],
    )


@router.get(
    "/sessions/{session_id}",
    summary="Get session details",
)
async def get_session(
    session_id: UUID,
    current_user: dict = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    """Get details of a collaborative editing session."""
    service = get_collaboration_service(db)
    session_data = await service._get_session_data(session_id)
    return session_data


@router.delete(
    "/sessions/{session_id}/leave",
    summary="Leave collaborative editing session",
)
async def leave_session(
    session_id: UUID,
    current_user: dict = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict[str, str]:
    """Leave a collaborative editing session."""
    user_id = UUID(current_user["user_id"])

    service = get_collaboration_service(db)
    await service.leave_session(session_id=session_id, user_id=user_id)

    return {"message": "Left session successfully"}


@router.patch(
    "/sessions/{session_id}/presence",
    summary="Update presence information",
)
async def update_presence(
    session_id: UUID,
    request: UpdatePresenceRequest,
    current_user: dict = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict[str, str]:
    """Update the current user's presence in a session.

    This includes cursor position, selected assets, and status.
    """
    user_id = UUID(current_user["user_id"])

    service = get_collaboration_service(db)
    await service.update_presence(
        session_id=session_id,
        user_id=user_id,
        status=request.status,
        cursor=request.cursor,
        selected_assets=request.selected_assets,
        focus_area=request.focus_area,
    )

    return {"message": "Presence updated"}


@router.post(
    "/sessions/{session_id}/actions",
    response_model=RecordActionResponse,
    summary="Record a collaborative action",
)
async def record_action(
    session_id: UUID,
    request: RecordActionRequest,
    current_user: dict = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> RecordActionResponse:
    """Record a collaborative action in the session.

    The action will be broadcast to all other participants.
    Conflict detection is performed if the base_version is stale.
    """
    user_id = UUID(current_user["user_id"])
    workspace_id = UUID(current_user["workspace_id"])

    # Get gallery_id from session
    service = get_collaboration_service(db)
    session_data = await service._get_session_data(session_id)
    gallery_id = UUID(session_data["gallery_id"])

    result = await service.record_action(
        session_id=session_id,
        user_id=user_id,
        workspace_id=workspace_id,
        gallery_id=gallery_id,
        action_type=request.action_type,
        payload=request.payload,
        base_version=request.base_version,
    )

    return RecordActionResponse(
        action_id=result["action_id"],
        success=result["success"],
        new_version=result["new_version"],
        conflict=result.get("conflict"),
    )


@router.get(
    "/sessions/{session_id}/actions",
    response_model=ActionHistoryResponse,
    summary="Get action history",
)
async def get_action_history(
    session_id: UUID,
    from_version: int = Query(0, description="Start version"),
    limit: int = Query(100, description="Max actions to return"),
    current_user: dict = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> ActionHistoryResponse:
    """Get the history of actions in a session.

    Useful for syncing state after reconnection.
    """
    service = get_collaboration_service(db)
    actions = await service.get_action_history(
        session_id=session_id,
        from_version=from_version,
        limit=limit,
    )

    to_version = actions[-1]["result_version"] if actions else from_version

    return ActionHistoryResponse(
        actions=actions,
        from_version=from_version,
        to_version=to_version,
    )


@router.post(
    "/sessions/{session_id}/actions/undo",
    summary="Undo an action",
)
async def undo_action(
    session_id: UUID,
    request: UndoActionRequest,
    current_user: dict = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    """Undo a previously recorded action."""
    user_id = UUID(current_user["user_id"])

    service = get_collaboration_service(db)
    result = await service.undo_action(
        session_id=session_id,
        user_id=user_id,
        action_id=request.action_id,
    )

    if not result["success"]:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=result.get("error", "Failed to undo action"),
        )

    return result


@router.post(
    "/sessions/{session_id}/locks",
    response_model=LockResponse,
    summary="Acquire a resource lock",
)
async def acquire_lock(
    session_id: UUID,
    request: AcquireLockRequest,
    current_user: dict = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> LockResponse:
    """Acquire a lock on a resource for exclusive or shared editing."""
    user_id = UUID(current_user["user_id"])
    workspace_id = UUID(current_user["workspace_id"])

    service = get_collaboration_service(db)
    result = await service.acquire_lock(
        workspace_id=workspace_id,
        session_id=session_id,
        user_id=user_id,
        resource_type=request.resource_type,
        resource_id=request.resource_id,
        lock_type=request.lock_type,
        duration_seconds=request.duration_seconds,
        reason=request.reason,
    )

    return LockResponse(
        success=result["success"],
        lock=result.get("lock"),
        error=result.get("error"),
        holder=result.get("holder"),
    )


@router.delete(
    "/locks/{lock_id}",
    summary="Release a resource lock",
)
async def release_lock(
    lock_id: UUID,
    current_user: dict = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    """Release a previously acquired lock."""
    user_id = UUID(current_user["user_id"])

    service = get_collaboration_service(db)
    result = await service.release_lock(lock_id=lock_id, user_id=user_id)

    if not result["success"]:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=result.get("error", "Failed to release lock"),
        )

    return result


# ---------------------------------------------------------------------------
# WebSocket Endpoint
# ---------------------------------------------------------------------------

@router.websocket("/ws/{session_id}")
async def collaboration_websocket(
    websocket: WebSocket,
    session_id: UUID,
    token: Annotated[str | None, Query()] = None,
) -> None:
    """WebSocket endpoint for real-time collaboration.

    Authentication:
    - Token from query parameter (obtained when creating/joining session)

    Message Types (incoming):
    - cursor: Update cursor position
    - selection: Update selected assets
    - action: Perform collaborative action
    - ping: Keepalive

    Message Types (outgoing):
    - collab:presence: Collaborator presence update
    - collab:cursor: Cursor position update
    - collab:action: Action broadcast
    - collab:lock_acquired/released: Lock events
    - collab:notification: Change notification
    - pong: Keepalive response
    """
    await websocket.accept()

    # Validate token
    if not token:
        await websocket.close(code=status.WS_1008_POLICY_VIOLATION)
        return

    redis = await get_redis_client()
    token_data = await redis.get(f"collab:token:{token}")
    if not token_data:
        await websocket.close(code=status.WS_1008_POLICY_VIOLATION)
        return

    token_info = json.loads(token_data)
    if token_info["session_id"] != str(session_id):
        await websocket.close(code=status.WS_1008_POLICY_VIOLATION)
        return

    user_id = token_info["user_id"]

    logger.info(
        f"WebSocket connected: user={user_id}, session={session_id}",
        extra={"user_id": user_id, "session_id": str(session_id)},
    )

    # Subscribe to Redis channel for this session
    pubsub = redis.pubsub()
    channel = f"{COLLAB_CHANNEL_PREFIX}{session_id}"
    await pubsub.subscribe(channel)

    try:
        # Send welcome message
        await websocket.send_json({
            "type": "collab:connected",
            "session_id": str(session_id),
            "user_id": user_id,
        })

        # Forward messages from Redis to WebSocket
        async def forward_messages():
            async for message in pubsub.listen():
                if message["type"] == "message":
                    try:
                        event_data = json.loads(message["data"])
                        # Don't send user's own cursor updates back
                        if (
                            event_data.get("type") == "collab:cursor"
                            and event_data.get("user_id") == user_id
                        ):
                            continue
                        await websocket.send_json(event_data)
                    except Exception as e:
                        logger.error(f"Failed to forward message: {e}")

        forward_task = asyncio.create_task(forward_messages())

        # Handle incoming WebSocket messages
        while True:
            try:
                data = await websocket.receive_text()

                if data == "ping":
                    await websocket.send_text("pong")
                    continue

                try:
                    message = json.loads(data)
                    message_type = message.get("type")

                    if message_type == "cursor":
                        # Broadcast cursor update
                        await redis.publish(
                            channel,
                            json.dumps({
                                "type": "collab:cursor",
                                "session_id": str(session_id),
                                "user_id": user_id,
                                "cursor": message.get("cursor"),
                            }),
                        )

                    elif message_type == "selection":
                        # Broadcast selection update
                        await redis.publish(
                            channel,
                            json.dumps({
                                "type": "collab:selection",
                                "session_id": str(session_id),
                                "user_id": user_id,
                                "selected_assets": message.get("selected_assets", []),
                            }),
                        )

                    elif message_type == "typing":
                        # Broadcast typing indicator
                        await redis.publish(
                            channel,
                            json.dumps({
                                "type": "collab:typing",
                                "session_id": str(session_id),
                                "user_id": user_id,
                                "field": message.get("field"),
                                "is_typing": message.get("is_typing", False),
                            }),
                        )

                except json.JSONDecodeError:
                    logger.warning(f"Invalid JSON message: {data[:100]}")

            except WebSocketDisconnect:
                break

    except Exception as e:
        logger.error(f"WebSocket error: {e}", exc_info=True)

    finally:
        # Cleanup
        forward_task.cancel()
        try:
            await forward_task
        except asyncio.CancelledError:
            pass

        await pubsub.unsubscribe(channel)
        await pubsub.close()

        logger.info(
            f"WebSocket disconnected: user={user_id}, session={session_id}",
            extra={"user_id": user_id, "session_id": str(session_id)},
        )

File: packages/shared-types/src/collaboration.ts
/**
 * Collaboration Types for Real-time Gallery Editing
 *
 * Provides type definitions for collaborative gallery editing features including:
 * - Edit sessions and presence tracking
 * - Real-time cursor positions
 * - Conflict resolution
 * - Change notifications
 */

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

/**
 * Status of an edit session
 */
export enum EditSessionStatus {
  ACTIVE = 'active',
  IDLE = 'idle',
  DISCONNECTED = 'disconnected',
  EXPIRED = 'expired',
}

/**
 * Type of collaborative action being performed
 */
export enum CollaborativeActionType {
  // Asset actions
  ASSET_ADD = 'asset:add',
  ASSET_REMOVE = 'asset:remove',
  ASSET_MOVE = 'asset:move',
  ASSET_REORDER = 'asset:reorder',
  ASSET_UPDATE = 'asset:update',
  ASSET_FAVORITE = 'asset:favorite',
  ASSET_SELECT = 'asset:select',

  // Collection/sub-gallery actions
  COLLECTION_CREATE = 'collection:create',
  COLLECTION_DELETE = 'collection:delete',
  COLLECTION_RENAME = 'collection:rename',
  COLLECTION_REORDER = 'collection:reorder',

  // Gallery metadata actions
  GALLERY_UPDATE = 'gallery:update',
  GALLERY_SETTINGS = 'gallery:settings',

  // AI suggestion actions
  AI_SUGGESTION_APPLY = 'ai:apply',
  AI_SUGGESTION_REJECT = 'ai:reject',
  AI_SUGGESTION_REQUEST = 'ai:request',

  // Cursor/selection actions
  CURSOR_MOVE = 'cursor:move',
  SELECTION_CHANGE = 'selection:change',
}

/**
 * Type of resource lock
 */
export enum LockType {
  EXCLUSIVE = 'exclusive',
  SHARED = 'shared',
}

/**
 * Conflict resolution strategy
 */
export enum ConflictResolutionStrategy {
  LAST_WRITE_WINS = 'last_write_wins',
  FIRST_WRITE_WINS = 'first_write_wins',
  MANUAL = 'manual',
  MERGE = 'merge',
}

/**
 * WebSocket message types for collaboration
 */
export enum CollaborationMessageType {
  // Connection lifecycle
  JOIN_SESSION = 'collab:join',
  LEAVE_SESSION = 'collab:leave',
  SESSION_JOINED = 'collab:session_joined',
  SESSION_LEFT = 'collab:session_left',

  // Presence
  PRESENCE_UPDATE = 'collab:presence',
  CURSOR_UPDATE = 'collab:cursor',
  SELECTION_UPDATE = 'collab:selection',

  // Actions
  ACTION_BROADCAST = 'collab:action',
  ACTION_ACK = 'collab:action_ack',
  ACTION_CONFLICT = 'collab:conflict',

  // Locking
  LOCK_REQUEST = 'collab:lock_request',
  LOCK_ACQUIRED = 'collab:lock_acquired',
  LOCK_RELEASED = 'collab:lock_released',
  LOCK_DENIED = 'collab:lock_denied',

  // Sync
  SYNC_STATE = 'collab:sync_state',
  SYNC_REQUEST = 'collab:sync_request',

  // Notifications
  CHANGE_NOTIFICATION = 'collab:notification',
  TYPING_INDICATOR = 'collab:typing',
}

// ---------------------------------------------------------------------------
// Core Interfaces
// ---------------------------------------------------------------------------

/**
 * Represents a user's presence in a collaborative session
 */
export interface CollaboratorPresence {
  /** User ID */
  user_id: string;
  /** Display name */
  display_name: string;
  /** Avatar URL (optional) */
  avatar_url?: string;
  /** Assigned color for visual identification */
  color: string;
  /** Current session status */
  status: EditSessionStatus;
  /** Current cursor position */
  cursor?: CursorPosition;
  /** Currently selected asset IDs */
  selected_assets: string[];
  /** Currently focused element/area */
  focus_area?: string;
  /** Last activity timestamp */
  last_activity: string;
  /** Session join timestamp */
  joined_at: string;
}

/**
 * Cursor position within the gallery editor
 */
export interface CursorPosition {
  /** X coordinate (normalized 0-1 or pixel position) */
  x: number;
  /** Y coordinate (normalized 0-1 or pixel position) */
  y: number;
  /** Target element ID or area */
  target?: string;
  /** Sub-gallery ID if in a sub-gallery view */
  sub_gallery_id?: string;
  /** Viewport scroll position */
  viewport?: {
    scroll_x: number;
    scroll_y: number;
    zoom: number;
  };
}

/**
 * Represents an edit session for a gallery
 */
export interface EditSession {
  /** Unique session ID */
  session_id: string;
  /** Gallery being edited */
  gallery_id: string;
  /** Workspace ID */
  workspace_id: string;
  /** Active collaborators */
  collaborators: CollaboratorPresence[];
  /** Session creation timestamp */
  created_at: string;
  /** Last activity in session */
  last_activity: string;
  /** Current version/revision number for conflict detection */
  version: number;
}

/**
 * Represents a collaborative action/operation
 */
export interface CollaborativeAction {
  /** Unique action ID */
  action_id: string;
  /** Type of action */
  action_type: CollaborativeActionType;
  /** User who performed the action */
  user_id: string;
  /** Session ID */
  session_id: string;
  /** Gallery ID */
  gallery_id: string;
  /** Action payload/data */
  payload: CollaborativeActionPayload;
  /** Version this action is based on */
  base_version: number;
  /** Timestamp of action */
  timestamp: string;
  /** Whether the action is undoable */
  undoable: boolean;
}

/**
 * Union type for action payloads
 */
export type CollaborativeActionPayload =
  | AssetAddPayload
  | AssetRemovePayload
  | AssetMovePayload
  | AssetReorderPayload
  | AssetUpdatePayload
  | CollectionCreatePayload
  | CollectionDeletePayload
  | CollectionRenamePayload
  | CollectionReorderPayload
  | GalleryUpdatePayload
  | AISuggestionPayload
  | CursorMovePayload
  | SelectionChangePayload;

// ---------------------------------------------------------------------------
// Action Payload Interfaces
// ---------------------------------------------------------------------------

export interface AssetAddPayload {
  asset_ids: string[];
  sub_gallery_id?: string;
  position?: number;
}

export interface AssetRemovePayload {
  asset_ids: string[];
}

export interface AssetMovePayload {
  asset_ids: string[];
  from_sub_gallery_id?: string;
  to_sub_gallery_id?: string;
}

export interface AssetReorderPayload {
  asset_ids: string[];
  sub_gallery_id?: string;
}

export interface AssetUpdatePayload {
  asset_id: string;
  updates: {
    title?: string;
    description?: string;
    tags?: string[];
    is_private?: boolean;
    is_favorited?: boolean;
    is_selected?: boolean;
  };
}

export interface CollectionCreatePayload {
  name: string;
  position?: number;
}

export interface CollectionDeletePayload {
  sub_gallery_id: string;
}

export interface CollectionRenamePayload {
  sub_gallery_id: string;
  name: string;
}

export interface CollectionReorderPayload {
  sub_gallery_ids: string[];
}

export interface GalleryUpdatePayload {
  updates: {
    title?: string;
    description?: string;
    settings?: Record<string, unknown>;
  };
}

export interface AISuggestionPayload {
  suggestion_id: string;
  suggestion_type: string;
  affected_assets?: string[];
  parameters?: Record<string, unknown>;
}

export interface CursorMovePayload {
  position: CursorPosition;
}

export interface SelectionChangePayload {
  asset_ids: string[];
  sub_gallery_id?: string;
}

// ---------------------------------------------------------------------------
// Lock Interfaces
// ---------------------------------------------------------------------------

/**
 * Represents a lock on a resource
 */
export interface ResourceLock {
  /** Lock ID */
  lock_id: string;
  /** Type of resource being locked */
  resource_type: 'gallery' | 'sub_gallery' | 'asset' | 'settings';
  /** Resource ID */
  resource_id: string;
  /** User holding the lock */
  user_id: string;
  /** Lock type */
  lock_type: LockType;
  /** Lock acquired timestamp */
  acquired_at: string;
  /** Lock expiry timestamp */
  expires_at: string;
  /** Optional reason for the lock */
  reason?: string;
}

/**
 * Request to acquire a lock
 */
export interface LockRequest {
  resource_type: 'gallery' | 'sub_gallery' | 'asset' | 'settings';
  resource_id: string;
  lock_type: LockType;
  duration_seconds?: number;
  reason?: string;
}

/**
 * Response to a lock request
 */
export interface LockResponse {
  success: boolean;
  lock?: ResourceLock;
  error?: string;
  holder?: {
    user_id: string;
    display_name: string;
  };
}

// ---------------------------------------------------------------------------
// Conflict Interfaces
// ---------------------------------------------------------------------------

/**
 * Represents a conflict between concurrent edits
 */
export interface EditConflict {
  /** Conflict ID */
  conflict_id: string;
  /** Conflicting actions */
  actions: [CollaborativeAction, CollaborativeAction];
  /** Type of conflict */
  conflict_type: 'concurrent_edit' | 'stale_version' | 'lock_violation';
  /** Affected resource */
  resource_type: string;
  resource_id: string;
  /** Suggested resolution */
  suggested_resolution: ConflictResolutionStrategy;
  /** Resolution options */
  resolution_options?: ConflictResolutionOption[];
  /** Timestamp of conflict detection */
  detected_at: string;
}

/**
 * Option for resolving a conflict
 */
export interface ConflictResolutionOption {
  /** Option ID */
  option_id: string;
  /** Description of the resolution */
  description: string;
  /** Preview of the result */
  preview?: unknown;
  /** Recommended option */
  recommended?: boolean;
}

/**
 * Request to resolve a conflict
 */
export interface ConflictResolutionRequest {
  conflict_id: string;
  resolution_option_id: string;
  /** Custom merge data if using MERGE strategy */
  merge_data?: unknown;
}

// ---------------------------------------------------------------------------
// WebSocket Message Interfaces
// ---------------------------------------------------------------------------

/**
 * Base WebSocket message
 */
export interface CollaborationMessage {
  type: CollaborationMessageType;
  session_id: string;
  timestamp: string;
}

/**
 * Join session request
 */
export interface JoinSessionMessage extends CollaborationMessage {
  type: CollaborationMessageType.JOIN_SESSION;
  gallery_id: string;
  user_id: string;
}

/**
 * Session joined response
 */
export interface SessionJoinedMessage extends CollaborationMessage {
  type: CollaborationMessageType.SESSION_JOINED;
  session: EditSession;
  your_color: string;
}

/**
 * Presence update message
 */
export interface PresenceUpdateMessage extends CollaborationMessage {
  type: CollaborationMessageType.PRESENCE_UPDATE;
  collaborator: CollaboratorPresence;
  event: 'joined' | 'left' | 'updated';
}

/**
 * Cursor update message
 */
export interface CursorUpdateMessage extends CollaborationMessage {
  type: CollaborationMessageType.CURSOR_UPDATE;
  user_id: string;
  cursor: CursorPosition;
}

/**
 * Action broadcast message
 */
export interface ActionBroadcastMessage extends CollaborationMessage {
  type: CollaborationMessageType.ACTION_BROADCAST;
  action: CollaborativeAction;
}

/**
 * Action acknowledgment message
 */
export interface ActionAckMessage extends CollaborationMessage {
  type: CollaborationMessageType.ACTION_ACK;
  action_id: string;
  success: boolean;
  new_version: number;
  error?: string;
}

/**
 * Conflict notification message
 */
export interface ConflictMessage extends CollaborationMessage {
  type: CollaborationMessageType.ACTION_CONFLICT;
  conflict: EditConflict;
}

/**
 * Change notification message
 */
export interface ChangeNotificationMessage extends CollaborationMessage {
  type: CollaborationMessageType.CHANGE_NOTIFICATION;
  user_id: string;
  user_name: string;
  action_type: CollaborativeActionType;
  summary: string;
  affected_items: string[];
}

// ---------------------------------------------------------------------------
// API Request/Response Interfaces
// ---------------------------------------------------------------------------

/**
 * Request to create an edit session
 */
export interface CreateEditSessionRequest {
  gallery_id: string;
}

/**
 * Response from creating an edit session
 */
export interface CreateEditSessionResponse {
  session: EditSession;
  websocket_url: string;
  token: string;
}

/**
 * Request to get session history
 */
export interface GetSessionHistoryRequest {
  session_id: string;
  from_version?: number;
  limit?: number;
}

/**
 * Response with session history
 */
export interface GetSessionHistoryResponse {
  actions: CollaborativeAction[];
  from_version: number;
  to_version: number;
}

/**
 * Request to undo an action
 */
export interface UndoActionRequest {
  action_id: string;
  session_id: string;
}

/**
 * Request to redo an action
 */
export interface RedoActionRequest {
  action_id: string;
  session_id: string;
}

// ---------------------------------------------------------------------------
// AI Suggestion Interfaces (for collaborative AI features)
// ---------------------------------------------------------------------------

/**
 * AI suggestion for gallery organization
 */
export interface AISuggestion {
  /** Suggestion ID */
  suggestion_id: string;
  /** Type of suggestion */
  suggestion_type: 'auto_organize' | 'smart_collection' | 'duplicate_detection' | 'quality_filter' | 'face_grouping';
  /** Description of the suggestion */
  description: string;
  /** Confidence score (0-1) */
  confidence: number;
  /** Preview data */
  preview: AISuggestionPreview;
  /** Estimated impact */
  impact: {
    assets_affected: number;
    collections_affected: number;
  };
  /** Actions to apply this suggestion */
  actions: CollaborativeAction[];
  /** Timestamp when suggestion was generated */
  generated_at: string;
  /** Expiry timestamp */
  expires_at: string;
}

/**
 * Preview of an AI suggestion
 */
export interface AISuggestionPreview {
  /** Before state (simplified) */
  before?: {
    collections: { id: string; name: string; asset_count: number }[];
    uncategorized_count: number;
  };
  /** After state (simplified) */
  after?: {
    collections: { id: string; name: string; asset_count: number }[];
    uncategorized_count: number;
  };
  /** Sample asset movements */
  sample_movements?: {
    asset_id: string;
    from_collection?: string;
    to_collection: string;
  }[];
}

/**
 * Request AI suggestions for a gallery
 */
export interface RequestAISuggestionsRequest {
  gallery_id: string;
  suggestion_types?: AISuggestion['suggestion_type'][];
}

/**
 * Response with AI suggestions
 */
export interface RequestAISuggestionsResponse {
  suggestions: AISuggestion[];
  processing_time_ms: number;
}
